### DISCUSSION TITLE
Run cypress tests in github actions, running up front-end and back-end http servers.

### DISCUSSION THREAD

**Discussion by alessandroAmedei** [community/community#25999] [Actions]
<p>Hi, i would like to run cypress tests on my Vue project linked to cypress dashboard in Github actions, triggered on push or pull events.<br>
To test my application, wich is written on NodeJS at the backend and Vue in the frontend, i have to run both https server in the github workflow. <br>
I am using cypress-io/github-action and it worked fine without running the backend server. This was my configuration:</p>
<pre><code>name: End-to-end tests
on: [push]
jobs:
  cypress-run:
    runs-on: windows-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v1

      - name: Cypress run
        uses: cypress-io/github-action@v1
        with:
          record: true
          start: npm run serve --prefix frontend/app
          wait-on: http://localhost:8080
        env:
          CYPRESS_RECORD_KEY: &lt;key&gt;
</code></pre>
<p>Now the goal is to have also a running server using:</p>
<pre><code>npm run start # --&gt; node index.js
</code></pre>
<p> So that i can get backend api listening on localhost:8000 (and vue in dev mode is refering to).</p>
<p>But i can not get that to work. I have tried various configuration like this: </p>
<pre><code>name: End-to-end tests
on: [push]
jobs:
  cypress-run:
    runs-on: windows-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v1

      - name: Run backend
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}
          run: npm run start

      - name: Cypress run
        uses: cypress-io/github-action@v1
        with:
          record: true
          start: npm run serve --prefix frontend/nix-app
          wait-on: http://localhost:8080
        env:
          # pass the Dashboard record key as an environment variable
          CYPRESS_RECORD_KEY: &lt;key&gt;
</code></pre>
<p>The backend server runs correctly on port 8000, but then the workflow is stuck there! It doesn’t continue on Cypress run because, i guess, it’s waiting the end of the process.<br>
How can i force to go at the next step after the first server is set up?</p>
<p>Another solution that i have tried is to use two different jobs in parallel, but also this is not working because the two servers will not stop running. Then what if cypress tests run before the two servers are up?</p>
<p>Can you suggest something else? Thank you!</p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/alessandroamedei">@alessandroamedei</a> , </p>
<p>To run a node.js app as a background service, you can start the Node JS server with <strong>nohup</strong> command as below.</p>
<pre><code>nobup node index.js
</code></pre>
<p> Have you tried this method? </p>
<p>If this could not work, It would be great if you could share a sample project here for me to investigate further. </p>

**Comment by alessandroAmedei** [comment]
<p>Hi <a class="mention" href="/yanjingzhu">@yanjingzhu</a> , thank you for the answer.</p>
<p>Now i have this:</p>
<pre><code>name: End-to-end tests
on: [push]
jobs:
  cypress-run:
    runs-on: windows-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v1

      - name: Set up node
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}
      - name: Run backend nohup
        run: nohup node backend/index.js

      - name: Cypress run
        uses: cypress-io/github-action@v1
        with:
          record: true
          start: npm run serve --prefix frontend/nix-app
          wait-on: http://localhost:8080
        env:
          # pass the Dashboard record key as an environment variable
          CYPRESS_RECORD_KEY: &lt;key&gt;
</code></pre>
<p>I am using two steps, i am doing :</p>
<pre><code>nohup node backend/index.js
</code></pre>
<p>But it is not working: </p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/dc9911e191041ba5939eeb4928af57a357d35f28.png" data-download-href="/uploads/short-url/vtuZDsq4EWb0ocoE1HMtrAjcQfC.png?dl=1" title="d1.PNG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/28606826/181094468-56ea3bab-74a1-47d8-812b-3de91262cbbc.png" alt="d1.PNG" data-base62-sha1="vtuZDsq4EWb0ocoE1HMtrAjcQfC" width="690" height="302" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/dc9911e191041ba5939eeb4928af57a357d35f28_2_690x302.png, https://aws1.discourse-cdn.com/github/optimized/2X/d/dc9911e191041ba5939eeb4928af57a357d35f28_2_1035x453.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/d/dc9911e191041ba5939eeb4928af57a357d35f28_2_1380x604.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/dc9911e191041ba5939eeb4928af57a357d35f28_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">d1.PNG</span><span class="informations">1829×803 44.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>It is still stuck.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>The <code>&amp;</code> symbol at the end of a command in bash will run the command in the background. </p>
<p>I tested it in powershell core on windows-latest runner, it could also work.</p>
<p>Please try to add &amp; at the end of the command which you want to run in backgroud . </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/f54b7999b3bbfded4546061edb34f8f5bf1f3312.png" data-download-href="/uploads/short-url/yZYGs0omiMHOcJFM9OeR71doT06.png?dl=1" title="backgroud job.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181094470-a3d8701c-f77a-4b4d-8e6b-54a6f2656fe0.png" alt="backgroud job.png" data-base62-sha1="yZYGs0omiMHOcJFM9OeR71doT06" width="690" height="289" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/f54b7999b3bbfded4546061edb34f8f5bf1f3312_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">backgroud job.png</span><span class="informations">933×392 7.32 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by JoeyChang123** [comment]
<p>Did you run the background process successfully?  Did you meet the this situation. At the last step, a job named Complete Job will clean up the orphan process. And this step will clean up the backgroup process unfortunately. I don’t know how do deal with this.</p>
